fideslib/db/base_class.py

```diff
@@ -0,0 +1,302 @@
+    created_at = Column(DateTime(timezone=True), server_default=func.now())
+    updated_at = Column(
+        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
+    )
+
+
+class OrmWrappedFidesBase(FidesBase):
+    """A wrapper class for our base model.
+
+    This allows us to include the base model in type checking for our return types.
+    This abstraction includes the following methods:
+    - get(id): return the record at that particular ID
+    - all(): return all records in that table
+    - filter(conditions): return all records that satisfy filter conditions
+    - create(data): create a record with provided data
+    - update_from_class(conditions, values): update all records that match the filter
+        conditions with the data inside values, called from the class
+    - update: update the record related to the object calling this method
+    - delete_with_class(id): delete the record at the provided ID, called from the class
+    - delete: delete the record related to the object calling this method
+    - delete_all: delete all records in this table
+    - save: update the record related to the object calling this method with the current
+        data stored on the object
+    """
+
+    __mapper_args__ = {"confirm_deleted_rows": False}
+    # This line hides sql alchemy warnings of the form:
+    # Sqlalchemy/orm/persistence.py:1461: SAWarning: DELETE statement on table
+    # 'storageconfig' expected to delete 1 row(s); 0 were matched. Please set
+    # confirm_deleted_rows=False within the mapper configuration to prevent this
+    # warning.
+
+    @classmethod
+    def get_optional_field_names(cls) -> list[str]:
+        """Returns the names of all nullable fields on the wrapped model"""
+        return [
+            field.name
+            for field in list(cls.__table__.columns)  # type: ignore # pylint: disable=E1101
+            if field.nullable
+        ]
+
+    @classmethod
+    def get(cls, db: Session, *, id: Any) -> FidesBase | None:  # pylint: disable=W0622
+        """Fetch a database record via a table ID"""
+        return db.query(cls).get(id)
+
+    @classmethod
+    def get_by(
+        cls,
+        db: Session,
+        *,
+        field: str,
+        value: str | int,
+    ) -> list[FidesBase]:
+        """Fetch a database record via a dynamic key, supplied at call time"""
+        kwargs = {field: value}
+        return db.query(cls).filter_by(**kwargs).first()  # type: ignore
+
+    @classmethod
+    def query(cls, db: Session) -> Query:
+        """Create a blank query for the class"""
+        return db.query(cls)
+
+    @classmethod
+    def all(cls, db: Session) -> list[OrmWrappedFidesBase]:  # pylint: disable=W0622
+        """Fetch all database records in table"""
+        return db.query(cls).all()
+
+    @classmethod
+    def filter(
+        cls,
+        db: Session,
+        *,
+        conditions: Any,  # TODO: What pydantic types are these? # pylint: disable=W0511
```

Comment:
Looks like it's a list of arguments: https://docs.sqlalchemy.org/en/14/orm/query.html?highlight=filter#sqlalchemy.orm.Query.filter

Reply:
Maybe a `**kwargs` argument, and we just document it in the docstring?

Reply:
it is an `*args`. Which now that you say that am wondering if this brings up a potential previously unseen bug.

what we effectively have is:

```py
def tester(*conditions):
    print(conditions)


def test(db, *, conditions):
    tester(conditions)

test(db="my_db", conditions=["condition1", "condition2"])
```

prints:

```console
(['condition1', 'condition2']) # actually prints this

('conditions1', 'conditions2')  # should print this
```

Should we force `conditions` to be passed as a list, then we can unpack it to send to `filter`?

```py
@classmethod
def update_with_class(
    cls, db: Session, *, conditions: list[Any], values: dict[str, Any]
) -> int:
    """Update all objects within a filter at database level"""
    return db.query(cls).filter(*conditions).update(values=values)
```

Reply:
I think `conditions` is best constrained to a `dict`, based on the sqlalchemy docs, no? That unpacking result _is_ unexpected though. I'm not sure 😕

Reply:
The way I am reading the docs (which to be honest aren't very clear here) is it is `*args`, not `**kwargs` so a `list` would be appropriate over `dict`. `dict` would be the choice if it was `**kwargs`.